[LETS-31] Add a class for all checkpoint information - PART 1 created a packable class

adrianbrab opened the pull request:
New checkpoint_info class was created. For the class 3 serialization functions have been implemented (pack, unpack, get_packed_size). Also an unit test with a couple of tested has been added to check that the packing is done correctly.

Commit:
added new checkpoint_info class

Commit:
added new checkpoint_info class

Commit:
added new checkpoint_info class

wrlawodms committed:
[CBRD-23866] Fix not to reuse heap when creating a table with REUSE_OID (#2584)

http://jira.cubrid.org/browse/CBRD-23866

airnet73 committed:
[CBRD-21446] Add environment variable and registry setting example file (#2594)

Commit:
added pack and unpack implementations

Commit:
added pack and unpack implementations

Commit:
added pack and unpack implementations

Commit:
Merge branch 'LETS-31_checkpoint_information_class' of github.com:adrianbrab/cubrid into LETS-31_checkpoint_information_class

Commit:
unit test started

Commit:
unit test started

Commit:
added random log_lsa generator

Commit:
added random log_lsa generator

Commit:
test building

Commit:
test building

Commit:
fixed segfault

Commit:
fixed segfault

Commit:
added new test cases

Commit:
added new test cases

ribram commented on cs/CMakeLists.txt:
```diff
@@ -292,6 +292,7 @@ set(TRANSACTION_SOURCES
   ${TRANSACTION_DIR}/locator.c
   ${TRANSACTION_DIR}/locator_cl.c
   ${TRANSACTION_DIR}/lock_table.c
+  ${TRANSACTION_DIR}/log_checkpoint_info.cpp
```
Not for CS mode, but it should be added to sa/CMakeLists.txt
```suggestion
```

ribram commented on cs/CMakeLists.txt:
```diff
@@ -301,6 +302,7 @@ set(TRANSACTION_SOURCES
 set(TRANSACTION_HEADERS
   ${TRANSACTION_DIR}/client_credentials.hpp
   ${TRANSACTION_DIR}/log_archives.hpp
+  ${TRANSACTION_DIR}/log_checkpoint_info.hpp
```
Not for CS mode, but it should be added to sa/CMakeLists.txt
```suggestion
```

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+
```
Add missing license header.

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+
+#include "log_checkpoint_info.hpp"
+namespace cublog
```
Leave an empty space after includes.
```suggestion
#include "log_checkpoint_info.hpp"

namespace cublog
```

ribram commented on src/transaction/log_checkpoint_info.hpp:
```diff
@@ -0,0 +1,44 @@
+      void set_start_redo_lsa (const log_lsa &start_redo_lsa);
+
+
+      log_lsa m_start_redo_lsa;
+      log_lsa m_snapshot_lsa;
+      std::vector<checkpoint_tran_info> m_trans;
+      std::vector<checkpoint_sysop_info> m_sysops;
+      bool m_has_2pc;				      // true if any LOG_ISTRAN_2PC (tdes) is true
```
The best practice is to make all members private.

ribram commented on src/transaction/log_record.hpp:
```diff
@@ -354,16 +354,108 @@ struct log_info_chkpt_trans
+	{
+	  return false;
+	}
+    }
```
Use std::strcmp from <cstring>
```suggestion
  if (std::strcmp (user_name, ochkpt.user_name) != 0)
    {
      return false;
    }
```

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+    deserializator.unpack_short  (offset);
+
+    return log_lsa (pageid, offset);
+  }
```
Best to keep blocks of code with different purposes separated. So the block with all checkpoint_definition should be separated from the block with the log_lsa packing helper functions

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+    m_snapshot_lsa = log_lsa_unpack (deserializator);
+
+    size_t m_trans_size = 0;
+    deserializator.unpack_bigint (m_trans_size);
```
- The m_ prefix is reserved for member names only, not local variables. Rename m_trans_size as trans_size.
- I am not sure [size_t ](https://en.cppreference.com/w/cpp/types/size_t)has 64 bits in all systems. To make sure that the correct type is passed to unpack_bigint, I suggest to change the type of trans_size to std::uint64_t:
```suggestion
    std::uint64_t trans_size = 0;
    deserializator.unpack_bigint (trans_size);
```

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+
+	int tran_state_int;
+	deserializator.unpack_int (tran_state_int);
+	chkpt_trans.state = static_cast<TRAN_STATE> (tran_state_int);
```
Use unpacker function unpack_from_int that automatically casts the into to the required type.
```suggestion
	deserializator.unpack_from_int (chkpt_trans.state);
```

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,159 @@
+	size = log_lsa_size (tran_info.savept_lsa, serializator, start_offset, size);
+	size = log_lsa_size (tran_info.tail_topresult_lsa, serializator, start_offset, size);
+	size = log_lsa_size (tran_info.start_postpone_lsa, serializator, start_offset, size);
+	size += serializator.get_packed_c_string_size (tran_info.user_name, strlen (tran_info.user_name), size);
```
```suggestion
	size += serializator.get_packed_c_string_size (tran_info.user_name, strlen (tran_info.user_name), start_offset + size);
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+#include <cstdlib>
+
+using namespace cublog;
+
+
```
Use only one empty line for separation.
```suggestion

```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+{
+  srand (time (0));
+
+  log_lsa log_to_add = this->generate_log_lsa();
+  before.m_start_redo_lsa = log_to_add;
```
Write the code in the easiest way possible:
```suggestion
  before.m_start_redo_lsa = this->generate_log_lsa();
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+  log_lsa log_to_add = this->generate_log_lsa();
+  before.m_start_redo_lsa = log_to_add;
+
+  log_to_add = this->generate_log_lsa();
+  before.m_snapshot_lsa = log_to_add;
```
Write the code in the easiest way possible:
```suggestion
  before.m_snapshot_lsa = this->generate_log_lsa();
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+{
+  LOG_INFO_CHKPT_TRANS chkpt_trans;
+
+  chkpt_trans.isloose_end = rand() % max;
```
```suggestion
  chkpt_trans.isloose_end = rand() % 2;  // can be true or false
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+
+  chkpt_trans.isloose_end = rand() % max;
+  chkpt_trans.trid        = rand() % max;
+  chkpt_trans.state       = static_cast<TRAN_STATE> (rand() % 15);
```
```suggestion
  chkpt_trans.state       = static_cast<TRAN_STATE> (rand() % TRAN_UNACTIVE_UNKNOWN);
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+    LOG_INFO_CHKPT_SYSOP generate_log_info_chkpt_sysop();
+    std::vector<LOG_LSA> used_logs;
+
+    const int max = 32700;
```
- Prefer constexpr instead of const for compile-time constants.
- Name constants using uppercase.
- Choose a name that makes purpose obvious.
```suggestion
    constexpr int MAX_RAND = 32700;
```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+{
+  LOG_LSA new_log = log_lsa (rand() % max, rand() % max);
+
+  while (std::find (used_logs.begin(), used_logs.end(), new_log) != used_logs.end())
+    {
+      new_log = log_lsa (rand() % max, rand() % max);
+    }
```
Is it necessary to completely avoid LSA duplicates? It seems overkill, random generation should be enough :).

adrianbrab replied:
I agree that it highly improbable for to generate the same the lsa, almost impossible for the structures. It looked like a good idea while writing it :), went back to the simpler faster generation

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,549 @@
+  REQUIRE (before.m_start_redo_lsa == after.m_start_redo_lsa);
+  REQUIRE (before.m_snapshot_lsa == after.m_snapshot_lsa);
+
+  REQUIRE (before.m_trans.size() == after.m_trans.size());
+  REQUIRE (before.m_trans == after.m_trans);
+
+  REQUIRE (before.m_sysops.size() == after.m_sysops.size());
+  REQUIRE (before.m_sysops == after.m_sysops);
```
Vector equality operator includes size compare.
```suggestion
  REQUIRE (before.m_trans == after.m_trans);

  REQUIRE (before.m_sysops == after.m_sysops);
```

adrianbrab replied:
They where added for debug purposes, removed now

Commit:
changed back to private

Commit:
changed back to private

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,550 @@
+#include <vector>
+#include <random>
+#include <cstdlib>
+
+
```
Remove duplicate empty line.
```suggestion

```

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,550 @@
+#include "catch2/catch.hpp"
+
+#define private public
+#include "log_checkpoint_info.hpp"
```
Undefine private immediately after the target include.
```suggestion
#include "log_checkpoint_info.hpp"
#undef private
```

Commit:
resolved comment suggestions

Commit:
resolved comment suggestions

kisoo-han committed:
[CBRD-23860] Do broker Keywords validation check (#2600)

* [CBRD-23860] Do broker Keywords validation check

ribram committed:
Merge remote-tracking branch 'cubrid/develop' into scalability_dev

ribram commented on src/transaction/log_checkpoint_info.cpp:
```diff
@@ -0,0 +1,175 @@
+	size = log_lsa_size (tran_info.savept_lsa, serializator, start_offset, size);
+	size = log_lsa_size (tran_info.tail_topresult_lsa, serializator, start_offset, size);
+	size = log_lsa_size (tran_info.start_postpone_lsa, serializator, start_offset, size);
+	size += serializator.get_packed_c_string_size (tran_info.user_name, strlen (tran_info.user_name),start_offset + size);
```
```suggestion
	size += serializator.get_packed_c_string_size (tran_info.user_name, strlen (tran_info.user_name), start_offset + size);
```

ribram commented on src/transaction/log_checkpoint_info.hpp:
```diff
@@ -0,0 +1,60 @@
+      void recovery_2pc_analysis () const;	      // if m_has_2pc, also do 2pc analysis
+
+      const log_lsa &get_snapshot_lsa () const;	      // the LSA of loaded snapshot
+      const log_lsa &get_start_redo_lsa ()
+      const;     // the LSA of starting redo. min of checkpoint LSA and oldest unflushed LSA.
```
```suggestion
      const log_lsa &get_start_redo_lsa () const;     // the LSA of starting redo. min of checkpoint LSA and oldest unflushed LSA.
```

ribram commented on src/transaction/log_record.hpp:
```diff
@@ -354,16 +354,106 @@ struct log_info_chkpt_trans
   LOG_LSA start_postpone_lsa;	/* Address of start postpone (if transaction was doing postpone during checkpoint) */
   char user_name[LOG_USERNAME_MAX];	/* Name of the client */
 
+
```
```suggestion

```

ribram commented on src/transaction/log_record.hpp:
```diff
@@ -354,16 +354,106 @@ struct log_info_chkpt_trans
+      return false;
+    }
+
+  if (std::strcmp (user_name, ochkpt.user_name) != 0)
```
Make sure <cstring> is included.

ribram commented on unit_tests/log_transfer/test_main_chkpt_info.cpp:
```diff
@@ -0,0 +1,531 @@
+
+test_env_chkpt::test_env_chkpt (int size_trans, int size_sysops)
+{
+  srand (time (0));
```
When using standard library functions, prefer using the std namespace:
```suggestion
  std::srand (time (0));
```
Same for all rand calls too.

Commit:
resolved code format problems

Commit:
resolved code format problems

ribram committed:
[LETS-47] Run scalability unit tests in circle-ci (#2604)

http://jira.cubrid.org/browse/LETS-47

Run the scalability unit tests in circle-ci.

  - Add a script that runs scalability unit tests.
  - Copy the script into the build folder
  - Modify circle-ci config file to compile unit tests and run the script during build phase.

The tests included in the script must be very small. No stress/performance tests will be allowed. Any longer unit tests should not be included or should be run in a short version.

Commit:
Merge branch 'LETS-31_checkpoint_information_class' of github.com:adrianbrab/cubrid into LETS-31_checkpoint_information_class

Commit:
added test to suite
